src/ir/auth_logic/datalog_ir.h

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright 2022 The Raksha Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// This file contains the datalog IR (DLIR) which makes the translation from
+// this authorization logic into datalog simpler.
+#ifndef SRC_IR_AUTH_LOGIC_DATALOG_IR_H_
+#define SRC_IR_AUTH_LOGIC_DATALOG_IR_H_
+
+#include <vector>
+
+#include "src/ir/auth_logic/ast.h"
+
+namespace raksha::ir::auth_logic {
+
+// A conditional datalog assertion with a left hand side and a right hand
+// side.
+class DLIRCondAssertion {
+ public:
+  explicit DLIRCondAssertion(Predicate lhs, std::vector<Predicate> rhs)
+      : lhs_(std::move(lhs)), rhs_(std::move(rhs)) {}
+
+ private:
+  Predicate lhs_;
+  std::vector<Predicate> rhs_;
+};
+
+// A Datalog IR assertion is either:
+//  - an unconditional fact which is just a predicate
+//  - a conditional assertion
+class DLIRAssertion {
+ public:
+  // DLIRAssertionVariantType represents the alternative forms for
+  // DLIRAssertions. Client code should use this type for traversing
+  // DLIRAssertions. This type may be changed in the future.
+  using DLIRAssertionVariantType = std::variant<Predicate, DLIRCondAssertion>;
```

Comment:
Same concerns as other uses of typedef-ed variants. OK for now as long as we resolve it soon.

Reply:
I'll resolve this one since we have another thread open to discuss this.